EIPS/eip-xxxx.md

```diff
@@ -0,0 +1,136 @@
+---
+eip: XXXX
+title: "Delayed Execution Layer State Root"
+description: "Delays the ExecutionPayload's state root reference by one block."
+author: Charlie Noyes <charlie@paradigm.xyz>, Dan Robinson <dan@paradigm.xyz>, Justin Drake <justin@ethereum.org>, Toni Wahrstätter <toni@ethereum.org>
+discussions-to: https://ethereum-magicians.org/t/eip-XXXX/, https://ethresear.ch/t/proposal-delay-stateroot-reference-to-increase-throughput-and-reduce-latency/20490
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-12-23
+---
+
+## Abstract
+
+This EIP proposes that, for each block `n`, the `state_root` in block `n`'s `ExecutionPayload` references the **post-state of block `n-1`**. As a result, block `n` no longer needs to include or verify the newly computed state root for block `n`’s own transactions. This removes the expensive state root computation from the block-propagation critical path, reducing end-to-end latency for block production and potentially increasing throughput.
+
+## Motivation
+
+Currently, every Ethereum block includes two `state_root`. 
+
+The consensus layer (CL) `state_root` contained within the `BeaconBlock` tracks CL state (e.g., validator entries and exits).
+
+The execution layer (EL) `state_root` contained within the `ExecutionPayload` tracks the effect of transaction execution (e.g., account balances, code, etc.). 
+
+The EL `state_root` must be computed by block builders (and verified by relays, validators, and light clients) in near-real-time. This computation accounts for a large fraction of block production and verification time, especially in an MEV-Boost environment. This overhead is also a challenge for efforts to enable real-time ZK proving of the chain.
+
+By delaying the EL `state_root` reference by one block, we can remove EL `state_root` calculation overhead from the critical path of block production. Instead, clients can pipeline calculation of block `n-1`'s EL `state_root` during the idle slot time of block `n`.
+
+This change will lower latency at the tip, allowing us to increase throughput and simplify the block production pipeline. It can also significantly accelerate the timeline real-time ZK proving Ethereum.
+
+## Specification
+
+### Execution Layer Changes
+
+1. **`ExecutionPayload` State Root Field**  
+   - **Before** (in the current post-Merge design):  
+     ```solidity
+     struct ExecutionPayload {
+       uint256 blockNumber;
+       uint256 timestamp;
+       bytes32 parentHash;
+       address feeRecipient;
+       ...
+       bytes32 stateRoot;  // references the post-state root of *this* block's transactions
+       ...
+     }
+     ```
+   - **After** (conceptual):  
+     ```solidity
+     struct ExecutionPayload {
+       uint256 blockNumber;
+       uint256 timestamp;
+       bytes32 parentHash;
+       address feeRecipient;
+       ...
+       bytes32 stateRoot;  // now references the post-state root of block (n-1)
```

Comment:
i have another Q: 

since the block builders will have the stateroot because they can't include a tx before executing a previous one because of previous txs affecting any prestate of the tx, so why should we not include the stateroot in it

since the logs/requests etc are all dependent on the correct execution. Ofcourse the attesters can still do delayed verification. 

may we we just add a pre-stateroot field because proofs of correct execution results against this block might require this block's post stateroot

Reply:
> since the block builders will have the stateroot because they can't include a tx before executing a previous one because of previous txs affecting any prestate of the tx, so why should we not include the stateroot in it

They only need the state, not the state root.

Currently they'd need to compute the state root before completing the block building process. Now they don't need to compute it eagerly (they can do it async).

Reply:
yes I get that, but having post stateroot in block is a good thing for creating proofs  merkle and ZK against the current block header itself.

yes it would matter how hard it is to create a stateroot, my supposition is : its not really that expensive and doesn't add any real latency but then depends on how big the block is, but if its vialble we should include it

Reply:
Sorry I'm not sure I understand. Do you suggest to include both pre state root and post state root of the current block in its header?

> having post stateroot in block is a good thing for creating proofs merkle and ZK against the current block header itself

Do you mean verifying blocks with a zk proof instead of re-executing them? That's not necessarily in conflict with this EIP, nodes can still receive the state root together with the proof (not as part of the header but on some other channel).

Reply:
yes I mean to specify both because the block builder will have the final execution results anyway when its done adding txs to the block.

for the proof, it needs to be anchored in post state root. Although receipts should have the entire execution traced but thats not true of system updates which are purely reflected in stateroot.

Reply:
ok now I understand that this EIP is trying to save post state root compute time even for builder. is that correct?

if so then as I asked in a previous comment how significant this time would be (so benchmarks against some typical blocks with 30m 60m... 100m 1g gas i guess) on a latest commondity hardware (which one would expect least of block builders)

I guess that would bring forward the proper justification of the EIP

Reply:
for purposes of accepting EIP draft just add a `<-- TODO benchmark -->` comment